lib/autotune/index.js

```diff
@@ -0,0 +1,292 @@
+            console.error(adjustmentRatio);
+            for (var offset=-3; offset < 0; offset++) {
+                offsetHour = hour + offset;
+                if (offsetHour < 0) { offsetHour += 24; }
+                hourlybasalprofile[offsetHour].rate = hourlybasalprofile[offsetHour].rate * adjustmentRatio;
+                hourlybasalprofile[offsetHour].rate=Math.round(hourlybasalprofile[offsetHour].rate*1000)/1000
+            }
+        }
+    }
+    if (pumpbasalprofile && pumpbasalprofile[0]) {
+        for (var hour=0; hour < 24; hour++) {
+            //console.error(hourlybasalprofile[hour],hourlypumpprofile[hour].rate*1.2);
+            // 20% caps
+            var maxrate = hourlypumpprofile[hour].rate * 1.2;
+            var minrate = hourlypumpprofile[hour].rate / 1.2;
+            if (hourlybasalprofile[hour].rate > maxrate ) {
+                console.error("Limiting hour",hour,"basal to",maxrate.toFixed(2),"(which is 20% above pump basal of",hourlypumpprofile[hour].rate,")");
+                hourlybasalprofile[hour].rate = maxrate;
+            } else if (hourlybasalprofile[hour].rate < minrate ) {
+                console.error("Limiting hour",hour,"basal to",minrate.toFixed(2),"(which is 20% below pump basal of",hourlypumpprofile[hour].rate,")");
+                hourlybasalprofile[hour].rate = minrate;
+            }
+            hourlybasalprofile[hour].rate = Math.round(hourlybasalprofile[hour].rate*1000)/1000;
+        }
+    }
+
+    console.error(hourlybasalprofile);
+    basalprofile = hourlybasalprofile;
+
+    // calculate median deviation and bgi in data attributable to ISF
+    var deviations = [];
+    var bgis = [];
+    var avgDeltas = [];
+    var ratios = [];
+    var count = 0;
+    for (var i=0; i < isf_glucose.length; ++i) {
+       deviation = parseFloat(isf_glucose[i].deviation);
+       deviations.push(deviation);
+       bgi = parseFloat(isf_glucose[i].bgi);
+       bgis.push(bgi);
+       avgDelta = parseFloat(isf_glucose[i].avgDelta);
+       avgDeltas.push(avgDelta);
+       ratio = 1 + deviation / bgi;
+       //console.error("Deviation:",deviation,"BGI:",bgi,"avgDelta:",avgDelta,"ratio:",ratio);
+       ratios.push(ratio);
+       count++;
+    }
+    avgDeltas.sort(function(a, b){return a-b});
+    bgis.sort(function(a, b){return a-b});
+    deviations.sort(function(a, b){return a-b});
+    ratios.sort(function(a, b){return a-b});
+    p50deviation = percentile(deviations, 0.50);
+    p50bgi = percentile(bgis, 0.50);
+    p50ratios = Math.round( percentile(ratios, 0.50) * 1000)/1000;
+
+    // calculate what adjustments to ISF would have been necessary to bring median deviation to zero
+    fullNewISF = isf * p50ratios;
+    fullNewISF = Math.round( fullNewISF * 1000 ) / 1000;
+    // and apply 10% of that adjustment
+    var newISF = ( 0.9 * isf ) + ( 0.1 * fullNewISF );
+    if (typeof(pumpISF) !== 'undefined') {
+        var maxISF = pumpISF * 1.2;
```

Comment:
Most people have been using 20% on ISF via autosens already, and we haven't had any issues with that - and watching the results of the tune, I feel comfortable continuing to have 20% be the safety cap for either one-off tuning or for use in the loop, given that it has the other safety parameters in concert with it, too. But yes, in the future we also plan to make it a ratio configurable in preferences.json, or to use the same ratio safety multipliers already there for autosens.